### PR TITLE
fixed crash with search extension

### PR DESCRIPTION
**Pull request by simonhong**
fix: https://github.com/brave/brave-browser/issues/15224
fix: https://github.com/brave/brave-browser/issues/10601
fix: https://github.com/brave/brave-browser/issues/14218

With this PR, brave uses search provider from extension by default on all platform like chromium does.

Crash comes from absense of exension search provider data in private profile's TemplateURLService because
we are using different TemplateURLService for private profile. When search provider extension is loaded/installed, browser doesn't notify about this adding to private profile's TUS.
To fix this, each SearchEngineProviderService sets extension's template url data to its service when normal profile's search provider comes from extension.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
`npm run test brave_browser_tests -- --filter=*ExtensionSearchProviderWithPrivateWindow*`

1. Launch browser
2. Install ecosia extension - https://chrome.google.com/webstore/detail/ecosia-the-search-engine/eedlgdlajadkbbjoobobefphmfkcchfk/related
3. Check Normal window's default search provider is ecosia
4. Open private window and check ecosia is default search provider
5. Turn ddg button on doesn't change search provider - still using ecosia
6. Turn ddg button off doesn't change search provider - still using ecosia
7. Goto ecosia extension details and toggle `Allow in private` repeatedly
8. Check changing `Allow in privatge` option doesn't affect private window's search provider - still using ecosia
9. Disable ecosia extension and check normal and private window uses default search provider
10. Toggle DDG button on in private NTP and check search provider is changed to DDG

### CODE REVIEW

**Comment by simonhong** [comment]
Hmm, it's really hard to control search provider by setting `kDefaultSearchProviderDataPrefName` manually whenever search provider extension's setting is changed.

File: browser/search_engines/tor_window_search_engine_provider_service.cc [hunk 1]
```diff
@@ -5,8 +5,11 @@
 
 #include "brave/browser/search_engines/tor_window_search_engine_provider_service.h"
 
+#include "brave/common/pref_names.h"
```
Comment (simonhong) [inline browser/search_engines/tor_window_search_engine_provider_service.cc:8]: delete this.

**Comment by simonhong** [comment]
Kindly ping :)

**Comment by iefremov** [comment]
I'll defer to @goodov

File: browser/search_engines/search_engine_provider_service.cc [hunk 2]
```diff
@@ -76,3 +85,47 @@ void SearchEngineProviderService::ChangeToNormalWindowSearchEngineProvider() {
   otr_template_url_service_->SetUserSelectedDefaultSearchProvider(
       &normal_url);
 }
+
+void SearchEngineProviderService::UseExtensionSearchProvider() {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  DCHECK(ShouldUseExtensionSearchProvider());
+
+  const auto* extension_provider_url =
```
Comment (goodov) [inline browser/search_engines/search_engine_provider_service.cc:93]: `DCHECK(extension_provider_url);`

File: browser/search_engines/search_engine_provider_service.h [hunk 3]
```diff
@@ -30,21 +35,22 @@ class SearchEngineProviderService : public KeyedService {
   void ChangeToAlternativeSearchEngineProvider();
   void ChangeToNormalWindowSearchEngineProvider();
 
+  bool ShouldUseExtensionSearchProvider();
```
Comment (goodov) [inline browser/search_engines/search_engine_provider_service.h:38]: `const`?

File: browser/search_engines/tor_window_search_engine_provider_service.cc [hunk 4]
```diff
@@ -15,16 +17,34 @@ TorWindowSearchEngineProviderService(Profile* otr_profile)
     : SearchEngineProviderService(otr_profile) {
   DCHECK(otr_profile->IsTor());
 
-  // Config default provider for tor window.
   auto provider_data = GetInitialSearchEngineProvider(otr_profile->GetPrefs());
-  TemplateURL provider_url(*provider_data);
-  otr_template_url_service_->SetUserSelectedDefaultSearchProvider(
-      &provider_url);
+  alternative_search_engine_url_for_tor_.reset(new TemplateURL(*provider_data));
```
Comment (goodov) [inline browser/search_engines/tor_window_search_engine_provider_service.cc:21]: `make_unique`

File: browser/search_engines/tor_window_search_engine_provider_service.cc [hunk 5]
```diff
@@ -15,16 +17,34 @@ TorWindowSearchEngineProviderService(Profile* otr_profile)
...
+
+  ConfigureSearchEngineProvider();
+
+  observation_.Observe(original_template_url_service_);
```
Comment (goodov) [inline browser/search_engines/tor_window_search_engine_provider_service.cc:25]: also should reset on `Shutdown`.

File: browser/search_engines/tor_window_search_engine_provider_service.h [hunk 6]
```diff
@@ -8,23 +8,34 @@
...
  private:
   std::unique_ptr<TemplateURLData> GetInitialSearchEngineProvider(
       PrefService* prefs) const;
+  void ConfigureSearchEngineProvider();
+
+  std::unique_ptr<TemplateURL> alternative_search_engine_url_for_tor_;
+  base::ScopedObservation<TemplateURLService, TemplateURLServiceObserver>
+      observation_{this};
 
   DISALLOW_COPY_AND_ASSIGN(TorWindowSearchEngineProviderService);
```
Comment (goodov) [inline browser/search_engines/tor_window_search_engine_provider_service.h:40]: we can also replace that with `= delete` and remove `#include macros.h`.

File: components/search_engines/pref_names.h [hunk 7]
```diff
@@ -0,0 +1,12 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Comment (goodov) [inline components/search_engines/pref_names.h:1]: what about extending `search_engines_pref_names.h`? less conflict.

File: patches/components-search_engines-default_search_manager.cc.patch [hunk 8]
```diff
@@ -0,0 +1,12 @@
+diff --git a/components/search_engines/default_search_manager.cc b/components/search_engines/default_search_manager.cc
+index 28c6f03ff7c22ad55c216407db1d303f63b99825..da7d13d678329a78f27b4ac050d82e81fcb7f93a 100644
+--- a/components/search_engines/default_search_manager.cc
++++ b/components/search_engines/default_search_manager.cc
+@@ -305,6 +305,7 @@ void DefaultSearchManager::LoadDefaultSearchEngineFromPrefs() {
+   // Check if default search preference is overriden by extension.
+   if (pref->IsExtensionControlled()) {
+     extension_default_search_ = std::move(turl_data);
++  LOADDEFAULTSEARCHENGINEFROMPREFS_BRAVE
```
Comment (goodov) [inline patches/components-search_engines-default_search_manager.cc.patch:9]: indent doesn't seem right.

File: browser/search_engines/search_engine_provider_service.cc [hunk 9]
```diff
@@ -76,3 +85,47 @@ void SearchEngineProviderService::ChangeToNormalWindowSearchEngineProvider() {
...
+
+bool SearchEngineProviderService::ShouldUseExtensionSearchProvider() {
+  return original_template_url_service_->IsExtensionControlledDefaultSearch();
+}
+
+bool SearchEngineProviderService::CouldAddExtensionTemplateURL(
+    const TemplateURL* url) {
+  DCHECK_NE(TemplateURL::NORMAL, url->type());
+  for (const auto* turl : otr_template_url_service_->GetTemplateURLs()) {
+    if (url->type() == turl->type() &&
```
Comment (goodov) [inline browser/search_engines/search_engine_provider_service.cc:126]: `DCHECK(url)` and `DCHECK(turl)` would be nice.

File: browser/search_engines/private_window_search_engine_provider_service.cc [hunk 10]
```diff
@@ -5,47 +5,63 @@
...
+  } else {
+    ConfigureSearchEngineProvider();
+  }
+
   // Monitor normal profile's search engine changing because private window
   // should that search engine provider when alternative search engine isn't
   // used.
-  original_template_url_service_->AddObserver(this);
-  ConfigureSearchEngineProvider();
+  observation_.Observe(original_template_url_service_);
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/search_engines/private_window_search_engine_provider_service.cc:32]: done.

File: browser/search_engines/search_engine_provider_service.cc (hunk 2 above)
Comment (simonhong) [reply browser/search_engines/search_engine_provider_service.cc:93]: done.

File: browser/search_engines/search_engine_provider_service.cc (hunk 9 above)
Comment (simonhong) [reply browser/search_engines/search_engine_provider_service.cc:129]: done.

File: browser/search_engines/search_engine_provider_service.h (hunk 3 above)
Comment (simonhong) [reply browser/search_engines/search_engine_provider_service.h]: done.

File: browser/search_engines/search_engine_provider_service.h [hunk 11]
```diff
@@ -30,21 +35,22 @@ class SearchEngineProviderService : public KeyedService {
...
+  TemplateURLService* otr_template_url_service_ = nullptr;
 
  private:
   void OnPreferenceChanged(const std::string& pref_name);
+  bool CouldAddExtensionTemplateURL(const TemplateURL* url);
 
+  std::unique_ptr<TemplateURL> alternative_search_engine_url_;
   BooleanPrefMember use_alternative_search_engine_provider_;
-
-  DISALLOW_COPY_AND_ASSIGN(SearchEngineProviderService);
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/search_engines/search_engine_provider_service.h:47]: done.

File: browser/search_engines/tor_window_search_engine_provider_service.h (hunk 6 above)
Comment (simonhong) [reply browser/search_engines/tor_window_search_engine_provider_service.h]: done.

File: components/search_engines/pref_names.h (hunk 7 above)
Comment (simonhong) [reply components/search_engines/pref_names.h]: done.

File: browser/search_engines/tor_window_search_engine_provider_service.cc (hunk 5 above)
Comment (simonhong) [reply browser/search_engines/tor_window_search_engine_provider_service.cc:26]: done.

File: browser/search_engines/tor_window_search_engine_provider_service.cc (hunk 4 above)
Comment (simonhong) [reply browser/search_engines/tor_window_search_engine_provider_service.cc]: done.

File: browser/search_engines/search_engine_provider_service.cc [hunk 12]
```diff
@@ -49,8 +59,7 @@ SearchEngineProviderService::SearchEngineProviderService(
   alternative_search_engine_url_.reset(new TemplateURL(*data));
 }
 
-SearchEngineProviderService::~SearchEngineProviderService() {
-}
+SearchEngineProviderService::~SearchEngineProviderService() = default;
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/search_engines/search_engine_provider_service.cc:62]: Hmm, I think we don't need to use it here.
This class uses one var (`BooleanPrefMember`) for monitoring prefs change, but I think it's fine to leave it to destroy in dtor.

**Review by simonhong — COMMENTED** [review]
@goodov Thanks for review. All addressed!

File: components/search_engines/search_engines_pref_names.h [hunk 13]
```diff
@@ -0,0 +1,12 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Comment (goodov) [inline components/search_engines/search_engines_pref_names.h:1]: I thought about extending chromium file via `chromium_src/components/search_engines/search_engines_pref_names.h`.
That way you don't need to add `brave_components_search_engines_sources` in `.gn`.

File: components/search_engines/pref_names.h (hunk 7 above)
Comment (goodov) [reply components/search_engines/pref_names.h]: I thought about extending chromium file via `chromium_src/components/search_engines/search_engines_pref_names.h`.
That way you don't need to add `brave_components_search_engines_sources` in `.gn`.

Comment (simonhong) [reply components/search_engines/pref_names.h]: @goodov  Oops! :) Addressed by https://github.com/brave/brave-core/pull/8626/commits/131c9aabd5e5bef4c60264b53107baac839dcab8

**Comment by goodov** [comment]
got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.

**Comment by simonhong** [comment]
@goodov Thanks for sharing DCHECK failure. Failure from `OnFaviconUpdated` is very new to me.
Will investigate this more. Thanks again! 👍

**Comment by simonhong** [comment]
> @simonhong
> got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
> reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.

@goodov  Fixed by https://github.com/brave/brave-core/pull/8626/commits/d091ba76b8d588a45ce34301e1cd715b2a07d809.  PTAL again!

**Comment by goodov** [comment]
> > @simonhong
> > got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
> > reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.
> 
> @goodov Fixed by [d091ba7](https://github.com/brave/brave-core/commit/d091ba76b8d588a45ce34301e1cd715b2a07d809). PTAL again!

Still doesn't seem right. After disabling the Extension search engine via Settings and enabling extension manually I got this error: https://gist.github.com/goodov/58a34a9ed19350984a9454b6522cae4d

**Comment by simonhong** [comment]
> > > @simonhong
> > > got this DCHECK while testing https://gist.github.com/goodov/5437d6735dce60bc24b9ef2943719c8e
> > > reproduced two times in private profile. after DuckDuck search via extension tried to search using Google by writing `:g test` in address bar.
> > 
> > 
> > @goodov Fixed by [d091ba7](https://github.com/brave/brave-core/commit/d091ba76b8d588a45ce34301e1cd715b2a07d809). PTAL again!
> 
> Still doesn't seem right. After disabling the Extension search engine via Settings and enabling extension manually I got this error: https://gist.github.com/goodov/58a34a9ed19350984a9454b6522cae4d

Fixed by https://github.com/brave/brave-core/pull/8626/files#diff-303c4f4e4e368bfb2681efc54ff35e1412d362740939fc50c7f93b9b92bcca2cR139. Changed to setting non-empty sync_guid. This PR will work :) Thanks for catching this.

And I'm trying another approach - This PR added new patches because we didn't make `kDefaultSearchProviderDataPrefName` as a extension controlled prefs. but I'm trying to set data to `kDefaultSearchProviderDataPrefName` as a extension controlled pref. If I do that, we can fix this issue w/o patching.

**Comment by goodov** [comment]
latest version seems fine, no any visible issues and the logic works great!

> And I'm trying another approach - This PR added new patches because we didn't make `kDefaultSearchProviderDataPrefName` as a extension controlled prefs. but I'm trying to set data to `kDefaultSearchProviderDataPrefName` as a extension controlled pref. If I can do that, we can fix this issue w/o patching.

so should we wait for solution without patching? :)

**Review by bridiver — APPROVED** [review]
approved, but please follow-up as discussed to see if we can get rid of the patch

**Comment by simonhong** [comment]
Merged because only unrelated brave ads/rewards browser tests were failed.

**Pull request merged**